### PR TITLE
[KILO]: DynComms Prefactor: Refactor/evaluate htlc view

### PR DESCRIPTION
**Pull request by ProofOfKeags**
## Change Description
This is the next logical chunk extracted off the front of #8755. This is a pure refactor change. The core behavior of the code modified should be identical.

## Steps to Test
make unit

### Testing
- [x] Your PR passes all CI checks.
- [x] Tests covering the positive and negative (error paths) are included.
- [x] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [x] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [x] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [x] Any new logging statements use an appropriate subsystem and logging level.
- [x] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [x] [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

### CODE REVIEW

File: lnwallet/channel.go
Comment (carlaKC) [inline lnwallet/channel.go:2782]: 😍 at last no `*ourBalance` monstery

File: lnwallet/channel.go
Comment (carlaKC) [inline lnwallet/channel.go:4365]: Not sure it would be better, but I'd be okay with a struct that expresses increase/decrease for readability:
```
type balanceDelta struct{
    balanceIncrease lnwire.Millisatoshi
    balanceDecrease lnwire.Millisatoshi
}
```

Could also have a `getBalance(currentBal)` method that does the addition + removal so that it's a single line per-party here.

**Review by carlaKC — COMMENTED** [review]
Nice refactor, simplifying power of `Dual` really shines 👌

File: lnwallet/channel.go
Comment (ProofOfKeags) [reply lnwallet/channel.go]: Seems like an improvement. I'll tack on a commit to the end.

File: lnwallet/channel.go
Comment (ProofOfKeags) [reply lnwallet/channel.go:2798]: Definitely. Thank you for giving me a nudge on it. I've been meaning to round out the API there.

File: lnwallet/channel.go
Comment (ProofOfKeags) [reply lnwallet/channel.go:2800]: Operationally this may be a change but fundamentally it is not. Irrespective of whether the update is locked in or not the feerate with respect to the view is *always* the last fee update issued by the channel initiator.

File: lnwallet/channel.go
Comment (ProofOfKeags) [reply lnwallet/channel.go:2782]: I will not rest until the demons have been excised. 🔪👻

File: lnwallet/channel.go
Comment (ProofOfKeags) [reply lnwallet/channel.go:4365]: In your imagined change here, would you allow both rows to be non-zero, or is this only a struct because go doesn't know what sum types are?

File: lnwallet/channel.go
Comment (carlaKC) [reply lnwallet/channel.go:2800]: > Irrespective of whether the update is locked in or not the feerate with respect to the view is always the last fee update issued by the channel initiator.

👍 this was what I thought too, glad to have somebody else confirm.

File: lnwallet/channel.go
Comment (carlaKC) [reply lnwallet/channel.go:4365]: Yeah, allow both to be non-zero when we sum up the deltas and then add/subtract each accordingly.
Just seems easier to not deal with negative numbers when we're around `lnwire.Millisatoshi` types.

File: lnwallet/channel.go
Comment (ProofOfKeags) [reply lnwallet/channel.go]: I remember why I didn't do this: 80col oppression

File: lnwallet/channel.go
Comment (ProofOfKeags) [reply lnwallet/channel.go]: So I went back and tried to do this and here's the issue. Notice how the thing we are filtering in the cited code is the uncommitted updates on `view`, *not* on `newView`. The process that occurs above this code specifically pairs off uncommitted settles and removes the corresponding add. In essence this evaluate view is a compression function and the compression loses the info we would need to subsequently compute the uncommitted settles. We *could* augment the view here to not lose that information so we can query it later, but I'm not thoroughly convinced that that's the right move. I'm open to input here.

The reason we need to keep the uncommitted settle information is so that we can properly record the lifetime sent/received accumulators in the view. If we don't have the uncommitted settles, we can't traverse them and accumulate their effects directly. Now we can *probably* repurpose the balance deltas to give us that information on. This is technically not semantics invariant the way the rest of the refactor series is, but it may be fine to do anyway. Lmk what you think here.

File: lnwallet/channel.go
Comment (ProofOfKeags) [reply lnwallet/channel.go:2798]: #9135 implements this.

File: lnwallet/channel.go
Comment (carlaKC) [reply lnwallet/channel.go]: >  80col oppression

tyranny ☠️ 

Barely looks better, but could go with:
```
return update.addCommitHeights.GetForParty(
   whoseCommitChain,
) == 0
```

File: lnwallet/channel.go
Comment (ProofOfKeags) [reply lnwallet/channel.go]: I can definitely put it as a method on the view, I just can't eliminate the need for the extra return value here since we need to report the pre-fold updates.

File: lnwallet/channel.go
Comment (ProofOfKeags) [reply lnwallet/channel.go:4600]: I think I'm going to punt on this change for now unless someone else chimes in saying they want it to work this way. Mostly I just think that having this structure might be more annoying then its worth. We can revisit later if it comes up again.

File: lnwallet/channel.go
Comment (carlaKC) [inline lnwallet/channel.go:4603]: I'd suggest pulling all of the code in this `if` out into a `LightningChannel.applyStateUpdate` method that's called here and in the test (tried in this [patch](https://github.com/user-attachments/files/17297552/update_state_closure.patch) and it's a minor lift).


We can leave a comment in the test that they used to be one function, and that the coverage should (eventually) be migrated to `computeView` _if_ it's not already covered.

File: lnwallet/channel.go
Comment (carlaKC) [inline lnwallet/channel.go:2904]: Comment on 2920 is outdated now that the mutation has been purged 🔪

File: lnwallet/channel_test.go
Comment (carlaKC) [inline lnwallet/channel_test.go:9200]: CI failing rn, but I would like to sanity check coveralls to make sure we're not losing any coverage before merge.

File: lnwallet/channel.go
Comment (carlaKC) [reply lnwallet/channel.go:4600]: Going to drop this commit completely then?

**Review by carlaKC — COMMENTED** [review]
<details>
    <summary>Nothing major left from me, will take a final pass when CI is happy + second reviewer has taken a look 👍</summary>

Screenshot 2024-10-08 at 2 26 44 PM


</details>

File: lnwallet/channel.go
Comment (ziggie1984) [inline lnwallet/channel.go:3053]: Nit: sent => set ?

File: lnwallet/channel.go
Comment (ziggie1984) [inline lnwallet/channel.go:4649]: Nit: [comment](https://github.com/lightningnetwork/lnd/pull/9097/files#diff-cdf9f47cef64ceac10ee55ed7bf0569ac54f5d5f1af599ef51ccaf5db813d03eR4658-R4661) needs a slight update, the underflow can happen now in computeView

File: lnwallet/channel.go
Comment (ziggie1984) [inline lnwallet/channel.go:2960]: why pointers here, we do not change the height in this part of the code ?

**Review by ziggie1984 — COMMENTED** [review]
Splendid refactor 🤩,

Approving when the CI passes.

Also I agree with Carla to not lose coverage here while removing some code parts.

**Review by yyforyongyu — COMMENTED** [review]
Most of my questions have been addressed offline, and just need to get the CI to pass.

File: lnwallet/channel.go
Comment (ProofOfKeags) [reply lnwallet/channel.go:3033]: Yes that's right. However I'd also offer a different way of looking at it:

- Adds only ever have the add height set.
- Settles/Fails only have the remove height set.
- FeeUpdates have both.
- The *definition* of "uncommitted" is not having the correct entry from the above list set.
- The point of returning the uncommitted updates is that we want to set the heights in the outer scope as opposed to threading the mutateState argument.
- When the next height gets locked in, everything that is uncommitted gets its heights set (for the given `whoseCommitChain`)

File: lnwallet/channel.go
Comment (ProofOfKeags) [reply lnwallet/channel.go:4649]: I don't know what comment is being referenced here. The link seems to go nowhere.

File: lnwallet/channel.go
Comment (ProofOfKeags) [reply lnwallet/channel.go:4603]: I can do this. However, I'm slightly nervous about it because if I make it a method, the bar for being well-defined is much higher.

This block appears to locking in the commit height of uncommitted payment descriptors *and* updating the lifetime payment flow statistics of the channel. I can't think of a coherent way to describe the *conjunction* of those things. What do you think about extracting out a more well-formed version of `setHeights` and `recordFlow` separately and calling them back to back here. Those at least are easier to give solid definitions and names for.

I mostly want to avoid attaching a function `applyStateUpdate` to `LightingChannel` for something so niche and incidental. If my suspicions are correct, you're asking me to do this to consolidate and reduce duplication of code that is fundamentally identical, not because it needs to be a singular function, and not also because it really should be a method on LightningChannel. Is that right?

File: lnwallet/channel.go
Comment (ProofOfKeags) [reply lnwallet/channel.go:2960]: Yep good catch.

File: lnwallet/channel.go
Comment (ProofOfKeags) [reply lnwallet/channel.go]: I figured out when to remove them. Should be correct now.

**Comment by ProofOfKeags** [comment]
CI is "as green as it gets". Coveralls unit is flat. All comments addressed. One possible TODO outstanding launched by a comment from @carlaKC. Re-requesting for final confirmation.

File: lnwallet/channel.go
Comment (carlaKC) [reply lnwallet/channel.go:4603]: > What do you think about extracting out a more well-formed version of setHeights and recordFlow separately and calling them back to back here. 

Happy with this! 

> If my suspicions are correct, you're asking me to do this to consolidate and reduce duplication of code that is fundamentally identical, not because it needs to be a singular function, and not also because it really should be a method on LightningChannel. Is that right?

Yes that's correct - don't like having the duplication just for the sake of not having to change the tests during this refactor.

Comment (ProofOfKeags) [reply lnwallet/channel.go:4603]: This change should be included in the most recent push. Lmk what you think.

**Review by carlaKC — APPROVED** [review]
Just reviewed the [push diff](https://github.com/lightningnetwork/lnd/compare/7e3cfe2cf972f0b177b029fbae4e25b06375f867..9bb9dd793b3e893ba595992a0687f5ee4ba4cefb), LGTM!

File: lnwallet/payment_descriptor.go
Comment (ziggie1984) [inline lnwallet/payment_descriptor.go:286]: nit: missing godoc

File: lnwallet/channel.go
Comment (ziggie1984) [inline lnwallet/channel.go:4669]: nit: missing godoc, hmm I wonder why the linter did not complain

File: lnwallet/channel.go
Comment (ziggie1984) [reply lnwallet/channel.go:4616]:
```
	// We need to first check ourBalance and theirBalance to be negative
	// because MilliSathoshi is a unsigned type and can underflow in
	// `evaluateHTLCView`. This should never happen for views which do not
	// include new updates (remote or local).
```
in `computeView` func

the underflow can now only happen in `computeView` itself because we do the `ourBalance` and `theirBalance` there now.

File: lnwallet/channel_test.go
Comment (ziggie1984) [inline lnwallet/channel_test.go:9197]: I understand that we do not need to check the mutate state anymore but where do we now test that the balances are updated correctly for the different payment descriptors ?

**Review by ziggie1984 — COMMENTED** [review]
Had a final question regarding the removed tests.

Can you rerun the `CI / check commits (pull_request)` test

File: lnwallet/channel_test.go
Comment (ziggie1984) [reply lnwallet/channel_test.go:9197]: Moreover the mutation is now pulled out into an upper level, are there unit-tests for it ? Cannot see a unit-test for `computeView` which now does the mutating.

File: lnwallet/channel.go
Comment (ProofOfKeags) [reply lnwallet/channel.go:4669]: linter only complains about godocs for exported functions, I think. I'll include anyway though.

File: lnwallet/channel_test.go
Comment (ProofOfKeags) [reply lnwallet/channel_test.go:9197]: as in the lifetime payment flows? I can try to design a test around that but I don't think the lifetime payment flow stuff is mission critical. With regards to the state updates of the payment descriptor commit heights, that is already implicitly tested by the itests during any payment that gets made, I think.

**Comment by ProofOfKeags** [comment]
> Can you rerun the CI / check commits (pull_request) test

It has been re-run and has passed.

File: lnwallet/channel_test.go
Comment (ziggie1984) [reply lnwallet/channel_test.go:9197]: ok normally I tend to favour having a complete unit-test suite rather than relying on itests, but given the fact that computeview had no unit-test beforehand having an unit-test for `computeView` would be nice but it's a non-blocking comment.

**Review by ziggie1984 — APPROVED** [review]

**Pull request merged**
